dwave/optimization/include/dwave-optimization/graph.hpp

```diff
@@ -293,21 +295,23 @@ NodeType* Graph::emplace_node(Args&&... args) {
     }
 
     // Construct via make_unique so we can allow the constructor to throw
-    nodes_.emplace_back(std::make_unique<NodeType>(std::forward<Args&&>(args)...));
-    NodeType* ptr = static_cast<NodeType*>(nodes_.back().get());
+    auto uptr = std::make_unique<NodeType>(std::forward<Args&&>(args)...);
+    NodeType* ptr = uptr.get();
+
+    // Pass ownership of the lifespan to nodes_
+    nodes_.emplace_back(std::move(uptr));
 
     // Decisions get their topological index assigned immediately, in the order of insertion
     if constexpr (std::is_base_of_v<Decision, NodeType>) {
-        ptr->topological_index_ = num_decisions_++;
+        static_assert(std::is_base_of_v<DecisionNode, NodeType>);
```

Comment:
This seems kind of redundant, no?

Reply:
It heads off someone accidentally inheriting from `Decision` but not `DecisionNode`. Like I did :laughing:

Reply:
Oops, missed the `Decision` vs `DecisionNode`! :+1: